idx/memory/tag_query.go

```diff
@@ -0,0 +1,411 @@
+package memory
+
+import (
+	"errors"
+	"math"
+	"regexp"
+
+	"github.com/grafana/metrictank/idx"
+	"github.com/raintank/worldping-api/pkg/log"
+)
+
+var (
+	errInvalidQuery = errors.New("invalid query")
+)
+
+const (
+	PARSING_ERROR = iota
+	EQUAL
+	NOT_EQUAL
+	MATCH
+	NOT_MATCH
+)
+
+type kv struct {
+	key   string
+	value string
+}
+
+type expression struct {
+	kv
+	operator int
+}
+
+type TagQuery struct {
+	from     int64
+	equal    []kv
+	match    []kv
+	notEqual []kv
+	notMatch []kv
+}
+
+// parseExpression returns an expression that's been generated from the given
+// string, in case of error the operator will be PARSING_ERROR.
```

Comment:
why not just return an `errInvalidQuery` error in that case, which is more go-ish

Reply:
done https://github.com/grafana/metrictank/pull/729/commits/92a850e3c381f78bbb4610d22880e8eb479988dd